docs/configuration.md
# ⚙ Configuration

## Settings

ReductStore uses environment variables for configuration. Here is a list of available settings:

| Name                | Default | Description                                                                               |
| ------------------- | ------- | ----------------------------------------------------------------------------------------- |
| RS\_LOG\_LEVEL      | INFO    | Logging level, can be: TRACE, DEBUG, INFO, WARNING, ERROR                                 |
| RS\_HOST            | 0.0.0.0 | Listening IP address                                                                      |
| RS\_PORT            | 8383    | Listening port                                                                            |
| RS\_API\_BASE\_PATH | /       | Prefix for all URLs of requests                                                           |
| RS\_DATA\_PATH      | /data   | Path to a folder where the storage stores the data                                        |
| RS\_API\_TOKEN      |         | If set, the storage uses [token authorization](broken-reference/)                         |
| RS\_CERT\_PATH      |         | Path to an SSL certificate. If unset, the storage uses HTTP instead of HTTPS              |
| RS\_CERT\_KEY\_PATH |         | Path to the private key of the desired SSL certificate. Should be set with RS\_CERT\_PATH |

### For Snap Users

If you use snap, you can configure the database by using the `snap set` command:

```
snap set reductstore log-level=DEBUG
```

This command changes the log level to DEBUG and restarts the database. You can check the current configuration with the `snap get reductstoret` command:

```
snap get reductstore
Key            Value
api-base       /
api-token
cert-key-path
cert-path
data-path      /var/snap/reductstore/common
host           0.0.0.0
log-level      DEBUG
port           8383
```

## Provisioning

ReductStore provides an HTTP API to create and configure resources such as buckets or access tokens. However, if you are following an Infrastructure as Code (IaC) approach, you may want to provision resources at the deployment stage and ensure that they can't be modified or deleted using the HTTP API. You can use the following environment variables to do this:

<table><thead><tr><th width="300">Name</th><th width="70">Default</th><th>Description</th></tr></thead><tbody><tr><td><strong>Bucket Provisioning</strong></td><td></td><td></td></tr><tr><td>RS_BUCKET_&#x3C;ID>_NAME</td><td></td><td>Provisioned bucket name (required)</td></tr><tr><td>RS_BUCKET_&#x3C;ID>_QUOTA_TYPE</td><td>NONE</td><td>It can be NONE or FIFO.  If it is FIFO, the bucket removes old data.</td></tr><tr><td>RS_BUCKET_&#x3C;ID>_QUOTA_SIZE</td><td>""</td><td>Size of quota to start removing old data e.g. 1 KB, 10.4 MB etc.</td></tr><tr><td>RS_BUCKET_&#x3C;ID>_MAX_BLOCK_SIZE</td><td>64Mb</td><td>Maximal block size for batched records</td></tr><tr><td>RS__BUCKET_&#x3C;ID>_MAX_BLOCK_RECORDS</td><td>256</td><td>Maximal number of batched records in a block</td></tr><tr><td><strong>Token Provisioning</strong></td><td></td><td></td></tr><tr><td>RS_TOKEN_&#x3C;ID>_NAME</td><td></td><td>Provisioned token name (required)</td></tr><tr><td>RS_TOKEN_&#x3C;ID>_VALUE</td><td></td><td>Provisioned value of token (required)</td></tr><tr><td>RS_TOKEN_&#x3C;ID>_FULL_ACCESS</td><td>false</td><td>Full access permission</td></tr><tr><td>RS_TOKEN_&#x3C;ID>_READ</td><td></td><td>List of buckets for reading</td></tr><tr><td>RS_TOKEN_&#x3C;ID>_WRITE</td><td></td><td>List of buckets for writing</td></tr></tbody></table>

{% hint style="info" %}
You can use any string value for \<ID>. It is only used to group resources of the same type.
{% endhint %}

&#x20;There is an example where we provide two buckets and a token to access them:

```
RS_BUCKET_A_NAME=bucket-1
RS_BUCKET_A_QUOTA_TYPE=FIFO
RS_BUCKET_A_QUOTA_SIZE=1Tb

RS_BUCKET_B_NAME=bucket-2

RS_TOKEN_A_NAME=token
RS_TOKEN_A_VALUE=somesecret
RS_TOKEN_A_READ=bucket-1,bucket-2
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

- reduct-cli: `alias` and `server` commands, [PR-343](https://github.com/reductstore/reductstore/pull/343)
- reduct-rs: `ReductClient.url`, `ReductClient.token`, `ReductCientBuilder.try_build` [PR-350](https://github.com/reductstore/reductstore/pull/350)
- reductstore: `healthcheck` to buildx.Dockerfile, [PR-350](https://github.com/reductstore/reductstore/pull/350)
- reductstore: provisioning with environment variables, [PR-352](https://github.com/reductstore/reductstore/pull/352)

### Changed

- reductstore: Update dependencies, min. rust v1.67.0, [PR-341](https://github.com/reductstore/reductstore/pull/341)
- reductstore: use Web Console v1.3.0, [PR-345](https://github.com/reductstore/reductstore/pull/342)
- reductstore: Move some Python API tests in Rust part, [PR-351](https://github.com/reductstore/reductstore/pull/351)
- reduct-base: Rename `HttpError` -> `ReductError`, [PR-350](https://github.com/reductstore/reductstore/pull/350)

### Fixed

- reduct-rs: Normalize instance URL in `ClientBuilder.url`, [PR-343](https://github.com/reductstore/reductstore/pull/343)

# [1.6.2] - 2023-09-20

### Fixed

- reductstore: Panic for a bad time interval in `GET /b/:bucket/:entry/q`, [PR-357](https://github.com/reductstore/reductstore/pull/357)

## [1.6.1] - 2023-08-28

### Fixed

- reductstore: README and LICENSE in reductstore crate, [PR-347](https://github.com/reductstore/reductstore/pull/347)

### Security

- reductstore: Update `rustls` with
  patched `rustls-webpki`, [PR-349](https://github.com/reductstore/reductstore/pull/349)

## [1.6.0] - 2023-08-14

### Added

- reductstore: Build docker image for ARM32 platform, [PR-328](https://github.com/reductstore/reductstore/pull/328)
- reductstore,reduct-rs: Removing entries from bucket, [PR-334](https://github.com/reductstore/reductstore/pull/334)
- reductstore,reduct-rs: Limit parameter in query request, [PR-335](https://github.com/reductstore/reductstore/pull/335)
- reduct-rs: Server API for Client SDK, [PR-321](https://github.com/reductstore/reductstore/pull/321)
- reduct-rs: Token API for Client SDK, [PR-322](https://github.com/reductstore/reductstore/pull/322)
- reduct-rs: Bucket API for Client SDK, [PR-323](https://github.com/reductstore/reductstore/pull/323)
- reduct-rs: Entry API for Client SDK, [PR-326](https://github.com/reductstore/reductstore/pull/326)
- reduct-rs: Examples and docs, [PR-333](https://github.com/reductstore/reductstore/pull/333)

### Changed

- reductstore: Refactor `http_frontend` module, [PR-306](https://github.com/reductstore/reductstore/pull/306)
- reductstore: Cache last block to reduce read operations, [PR-318](https://github.com/reductstore/reductstore/pull/318)
- reductstore: Grained HTTP components, [PR-319](https://github.com/reductstore/reductstore/pull/319)
- reductstore: Default maximum records in block 256, [PR-320](https://github.com/reductstore/reductstore/pull/320)
- reductstore: BUSL-1.1 license, [PR-337](https://github.com/reductstore/reductstore/pull/337)
- reductstore: Update README.md, [PR-338](https://github.com/reductstore/reductstore/pull/338)
- all: Organize workspaces, [PR-310](https://github.com/reductstore/reductstore/pull/310)

### Removed

- reductstore: `native-tls` dependency (only `rustls`), [PR-315](https://github.com/reductstore/reductstore/pull/315)

### Fixed

- reductstore: Partial bucket settings, [PR-325](https://github.com/reductstore/reductstore/pull/325)

## [1.5.1] - 2023-07-17

### Fixed

- Handle empty or broken block descriptor, [PR-317](https://github.com/reductstore/reductstore/pull/317)

## [1.5.0] - 2023-06-30

### Added

- `x-reduct-api` header to get quick version number in Major.Minor
  format, [PR-291](https://github.com/reductstore/reductstore/pull/291)
- `GET /api/v1/:bucket/:entry/batch` endpoint to read a bunch of
  records, [PR-294](https://github.com/reductstore/reductstore/pull/294)
- `HEAD /api/v1/b/:bucket_name/:entry_name` and `HEAD /api/v1/b/:bucket_name/:entry_name/batch`
  endpoints, [PR-296]https://github.com/reductstore/reductstore/pull/296)

### Changed

- Concise format for headers in `GET /api/v1/:bucket/:entry/batch`
  response, [PR-298](https://github.com/reductstore/reductstore/pull/298)

## [1.4.1] - 2023-06-27

### Fixed

- Stuck empty entries, [PR-302](https://github.com/reductstore/reductstore/pull/302)

## [1.4.0] - 2023-06-09

### Fixed

- Panic when an invalid utf-8 received as a label value, [PR-290](https://github.com/reductstore/reductstore/pull/290)
- Writing record for clients which don't support for Expect
  header, [PR-293](https://github.com/reductstore/reductstore/pull/293)

## [1.4.0-beta.1] - 2023-06-03

### Changed

- Update web console to 1.2.2, [PR-287](https://github.com/reductstore/reductstore/pull/287)

### Fixed

- Parsing non-string quota type, [PR-286](https://github.com/reductstore/reductstore/pull/286)
- Show file paths in logs without registry path, [PR-288](https://github.com/reductstore/reductstore/pull/288)

## [1.4.0-alpha.3] - 2023-05-29

### Fixed

- Return `init-token` to token list, [PR-280](https://github.com/reductstore/reductstore/pull/280)
- First query ID is 1, [PR-281](https://github.com/reductstore/reductstore/pull/281)
- Showing permissions in `GET /api/v1/tokens`, [PR-282](https://github.com/reductstore/reductstore/pull/282)
- Remove removed bucket from token permissions, [PR-283](https://github.com/reductstore/reductstore/pull/283)
- Build on Windows and macOS, [PR-284](https://github.com/reductstore/reductstore/pull/284)

## [1.4.0-alpha.2] - 2023-05-26


reductstore/src/cfg.rs
// Copyright 2023 ReductStore
// Licensed under the Business Source License 1.1

use crate::api::Components;
use crate::asset::asset_manager::ZipAssetManager;
use crate::auth::token_auth::TokenAuthorization;
use crate::auth::token_repository::{create_token_repository, ManageTokens};
use crate::core::env::{Env, GetEnv};

use crate::storage::storage::Storage;
use bytesize::ByteSize;

use log::{error, info, warn};
use reduct_base::error::ErrorCode;
use reduct_base::msg::bucket_api::BucketSettings;
use reduct_base::msg::token_api::{Permissions, Token};
use std::collections::HashMap;
use std::fmt::{Display, Formatter};
use std::path::PathBuf;

use tokio::sync::RwLock;

/// Database configuration
pub struct Cfg<EnvGetter: GetEnv> {
    pub log_level: String,
    pub host: String,
    pub port: i32,
    pub api_base_path: String,
    pub data_path: String,
    pub api_token: String,
    pub cert_path: String,
    pub cert_key_path: String,
    pub buckets: HashMap<String, BucketSettings>,
    pub tokens: HashMap<String, Token>,

    env: Env<EnvGetter>,
}

impl<EnvGetter: GetEnv> Cfg<EnvGetter> {
    pub fn from_env(env_getter: EnvGetter) -> Self {
        let mut env = Env::new(env_getter);
        let cfg = Cfg {
            log_level: env.get("RS_LOG_LEVEL", "INFO".to_string()),
            host: env.get("RS_HOST", "0.0.0.0".to_string()),
            port: env.get("RS_PORT", 8383),
            api_base_path: env.get("RS_API_BASE_PATH", "/".to_string()),
            data_path: env.get("RS_DATA_PATH", "/data".to_string()),
            api_token: env.get_masked("RS_API_TOKEN", "".to_string()),
            cert_path: env.get_masked("RS_CERT_PATH", "".to_string()),
            cert_key_path: env.get_masked("RS_CERT_KEY_PATH", "".to_string()),
            buckets: Self::parse_buckets(&mut env),
            tokens: Self::parse_tokens(&mut env),

            env,
        };

        cfg
    }

    pub fn build(&self) -> Components {
        let storage = self.provision_storage();
        let token_repo = self.provision_tokens();

        Components {
            storage: RwLock::new(storage),
            token_repo: RwLock::new(token_repo),
            auth: TokenAuthorization::new(&self.api_token),
            console: ZipAssetManager::new(include_bytes!(concat!(env!("OUT_DIR"), "/console.zip"))),
            base_path: self.api_base_path.clone(),
        }
    }

    fn provision_tokens(&self) -> Box<dyn ManageTokens + Send + Sync> {
        let mut token_repo =
            create_token_repository(PathBuf::from(self.data_path.clone()), &self.api_token);

        for (name, token) in &self.tokens {
            let is_generated = match token_repo
                .generate_token(&name, token.permissions.clone().unwrap_or_default())
            {
                Ok(_) => Ok(()),
                Err(e) => {
                    if e.status() == ErrorCode::Conflict {
                        Ok(())
                    } else {
                        Err(e)
                    }
                }
            };

            if let Err(err) = is_generated {
                error!("Failed to provision token '{}': {}", name, err);
            } else {
                let update_token = token_repo.get_mut_token(&name).unwrap();
                update_token.clone_from(token);
                update_token.is_provisioned = true;

                info!(
                    "Provisioned token '{}' with {:?}",
                    update_token.name, update_token.permissions
                );
            }
        }
        token_repo
    }

    fn provision_storage(&self) -> Storage {
        let mut storage = Storage::new(PathBuf::from(self.data_path.clone()));
        for (name, settings) in &self.buckets {
            let settings = match storage.create_bucket(&name, settings.clone()) {
                Ok(bucket) => {
                    bucket.set_provisioned(true);
                    Ok(bucket.settings().clone())
                }
                Err(e) => {
                    if e.status() == ErrorCode::Conflict {
                        let bucket = storage.get_mut_bucket(&name).unwrap();
                        bucket.set_provisioned(false);
                        bucket.set_settings(settings.clone()).unwrap();
                        bucket.set_provisioned(true);

                        Ok(bucket.settings().clone())
                    } else {
                        Err(e)
                    }
                }
            };

            if let Ok(settings) = settings {
                info!("Provisioned bucket '{}' with: {:?}", name, settings);
            } else {
                error!(
                    "Failed to provision bucket '{}': {}",
                    name,
                    settings.err().unwrap()
                );
            }
        }
        storage
    }

    fn parse_buckets(env: &mut Env<EnvGetter>) -> HashMap<String, BucketSettings> {
        let mut buckets = HashMap::<String, (String, BucketSettings)>::new();
        for (id, name) in env.matches("RS_BUCKET_(.*)_NAME") {
            buckets.insert(id, (name, BucketSettings::default()));
        }

        for (id, bucket) in &mut buckets {
            let settings = &mut bucket.1;
            settings.quota_type = env.get_optional(&format!("RS_BUCKET_{}_QUOTA_TYPE", id));

            settings.quota_size = env
                .get_optional::<ByteSize>(&format!("RS_BUCKET_{}_QUOTA_SIZE", id))
                .map(|s| s.as_u64());
            settings.max_block_size = env
                .get_optional::<ByteSize>(&format!("RS_BUCKET_{}_MAX_BLOCK_SIZE", id))
                .map(|s| s.as_u64());
            settings.max_block_records =
                env.get_optional(&format!("RS_BUCKET_{}_MAX_BLOCK_RECORDS", id));
        }

        buckets
            .into_iter()
            .map(|(_id, (name, settings))| (name, settings))
            .collect()
    }

    fn parse_tokens(env: &mut Env<EnvGetter>) -> HashMap<String, Token> {
        let mut tokens = HashMap::<String, Token>::new();
        for (id, name) in env.matches("RS_TOKEN_(.*)_NAME") {
            let token = Token {
                name,
                created_at: chrono::Utc::now(),
                ..Token::default()
            };
            tokens.insert(id, token);
        }

        for (id, token) in &mut tokens {
            token.value =
                env.get_masked::<String>(&format!("RS_TOKEN_{}_VALUE", id), "".to_string());
        }

        tokens.retain(|_, token| {
            if token.value.is_empty() {
                warn!("Token '{}' has no value. Drop it.", token.name);
                false
            } else {
                true
            }
        });

        let parse_list_env = |env: &mut Env<EnvGetter>, name: String| -> Vec<String> {
            env.get_optional::<String>(&name)
                .unwrap_or_default()
                .split(",")
                .map(|s| s.to_string())
                .collect()
        };

        // Parse permissions
        for (id, token) in &mut tokens {
            let read = parse_list_env(env, format!("RS_TOKEN_{}_READ", id));
            let write = parse_list_env(env, format!("RS_TOKEN_{}_WRITE", id));
            let permissions = Permissions {
                full_access: env
                    .get_optional(&format!("RS_TOKEN_{}_FULL_ACCESS", id))
                    .unwrap_or_default(),
                read,
                write,
            };

            token.permissions = Some(permissions);
        }

        tokens
            .into_iter()
            .map(|(_, token)| (token.name.clone(), token))
            .collect()
    }
}

impl<EnvGetter: GetEnv> Display for Cfg<EnvGetter> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.env.message())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use mockall::mock;
    use mockall::predicate::eq;
    use rstest::{fixture, rstest};
    use std::collections::BTreeMap;
    use std::env::VarError;

    mock! {
        EnvGetter {}
        impl GetEnv for EnvGetter {
            fn get(&self, key: &str) -> Result<String, VarError>;
            fn all(&self) -> BTreeMap<String,String>;
        }
    }

    #[rstest]
    fn test_default_settings(mut env_getter: MockEnvGetter) {
        env_getter
            .expect_get()
            .return_const(Err(VarError::NotPresent));

        let cfg = Cfg::from_env(env_getter);
        assert_eq!(cfg.log_level, "INFO");
        assert_eq!(cfg.host, "0.0.0.0");
        assert_eq!(cfg.port, 8383);
        assert_eq!(cfg.api_base_path, "/");
        assert_eq!(cfg.data_path, "/data");
        assert_eq!(cfg.api_token, "");
        assert_eq!(cfg.cert_path, "");
        assert_eq!(cfg.cert_key_path, "");

        assert_eq!(cfg.buckets.len(), 0);
        assert_eq!(cfg.tokens.len(), 0);
    }

    #[rstest]
    fn test_log_level(mut env_getter: MockEnvGetter) {
        env_getter
            .expect_get()
            .with(eq("RS_LOG_LEVEL"))
            .times(1)
            .return_const(Ok("DEBUG".to_string()));
        env_getter
            .expect_get()
            .return_const(Err(VarError::NotPresent));
        let cfg = Cfg::from_env(env_getter);
        assert_eq!(cfg.log_level, "DEBUG");
    }

    #[rstest]
    fn test_host(mut env_getter: MockEnvGetter) {
        env_getter
            .expect_get()
            .with(eq("RS_HOST"))
            .times(1)
            .return_const(Ok("127.0.0.1".to_string()));
        env_getter
            .expect_get()
            .return_const(Err(VarError::NotPresent));
        let cfg = Cfg::from_env(env_getter);
        assert_eq!(cfg.host, "127.0.0.1");
    }

    #[rstest]
    fn test_port(mut env_getter: MockEnvGetter) {
        env_getter
            .expect_get()
            .with(eq("RS_PORT"))
            .times(1)
            .return_const(Ok("1234".to_string()));
        env_getter
            .expect_get()
            .return_const(Err(VarError::NotPresent));
        let cfg = Cfg::from_env(env_getter);
        assert_eq!(cfg.port, 1234);
    }

    #[rstest]
    fn test_api_base_path(mut env_getter: MockEnvGetter) {
        env_getter
            .expect_get()
            .with(eq("RS_API_BASE_PATH"))
            .times(1)
            .return_const(Ok("/api".to_string()));
        env_getter
            .expect_get()
            .return_const(Err(VarError::NotPresent));
        let cfg = Cfg::from_env(env_getter);
        assert_eq!(cfg.api_base_path, "/api");
    }

    #[rstest]
    fn test_data_path(mut env_getter: MockEnvGetter) {
        env_getter
            .expect_get()
            .with(eq("RS_DATA_PATH"))
            .times(1)
            .return_const(Ok("/tmp".to_string()));
        env_getter
            .expect_get()
            .return_const(Err(VarError::NotPresent));
        let cfg = Cfg::from_env(env_getter);
        assert_eq!(cfg.data_path, "/tmp");
    }

    #[rstest]
    fn test_api_token(mut env_getter: MockEnvGetter) {
        env_getter
            .expect_get()
            .with(eq("RS_API_TOKEN"))
            .times(1)
            .return_const(Ok("XXX".to_string()));
        env_getter
            .expect_get()
            .return_const(Err(VarError::NotPresent));
        let cfg = Cfg::from_env(env_getter);
        assert_eq!(cfg.api_token, "XXX");
    }

    #[rstest]
    fn test_cert_path(mut env_getter: MockEnvGetter) {
        env_getter
            .expect_get()
            .with(eq("RS_CERT_PATH"))
            .times(1)
            .return_const(Ok("/tmp/cert.pem".to_string()));
        env_getter
            .expect_get()
            .return_const(Err(VarError::NotPresent));
        let cfg = Cfg::from_env(env_getter);
        assert_eq!(cfg.cert_path, "/tmp/cert.pem");
    }

    #[rstest]
    fn test_cert_key_path(mut env_getter: MockEnvGetter) {
        env_getter
            .expect_get()
            .with(eq("RS_CERT_KEY_PATH"))
            .times(1)
            .return_const(Ok("/tmp/cert.key".to_string()));
        env_getter
            .expect_get()
            .return_const(Err(VarError::NotPresent));
        let cfg = Cfg::from_env(env_getter);
        assert_eq!(cfg.cert_key_path, "/tmp/cert.key");
    }

    #[fixture]
    fn env_getter() -> MockEnvGetter {
        let mut mock_getter = MockEnvGetter::new();
        mock_getter.expect_all().returning(|| BTreeMap::new());
        return mock_getter;
    }

    #[fixture]
    fn env_with_buckets() -> MockEnvGetter {
        let tmp = tempfile::tempdir().unwrap();
        let mut mock_getter = MockEnvGetter::new();
        mock_getter
            .expect_get()
            .with(eq("RS_DATA_PATH"))
            .return_const(Ok(tmp.into_path().to_str().unwrap().to_string()));
        mock_getter.expect_all().returning(|| {
            let mut map = BTreeMap::new();
            map.insert("RS_BUCKET_1_NAME".to_string(), "bucket1".to_string());
            map
        });
        mock_getter
            .expect_get()
            .with(eq("RS_BUCKET_1_NAME"))
            .return_const(Ok("bucket1".to_string()));
        mock_getter
    }

    #[fixture]
    fn env_with_tokens() -> MockEnvGetter {
        let tmp = tempfile::tempdir().unwrap();
        let mut mock_getter = MockEnvGetter::new();
        mock_getter
            .expect_get()
            .with(eq("RS_DATA_PATH"))
            .return_const(Ok(tmp.into_path().to_str().unwrap().to_string()));
        mock_getter
            .expect_get()
            .with(eq("RS_API_TOKEN"))
            .return_const(Ok("XXX".to_string()));
        mock_getter.expect_all().returning(|| {
            let mut map = BTreeMap::new();
            map.insert("RS_TOKEN_1_NAME".to_string(), "token1".to_string());
            map
        });
        mock_getter
            .expect_get()
            .with(eq("RS_TOKEN_1_NAME"))
            .return_const(Ok("token1".to_string()));
        mock_getter
    }

    mod provision {
        use super::*;
        use crate::storage::bucket::Bucket;
        use reduct_base::error::ReductError;
        use reduct_base::msg::bucket_api::QuotaType::FIFO;

        #[rstest]
        #[tokio::test]
        async fn test_buckets(mut env_with_buckets: MockEnvGetter) {
            env_with_buckets
                .expect_get()
                .with(eq("RS_BUCKET_1_QUOTA_TYPE"))
                .return_const(Ok("FIFO".to_string()));
            env_with_buckets
                .expect_get()
                .with(eq("RS_BUCKET_1_QUOTA_SIZE"))
                .return_const(Ok("1GB".to_string()));
            env_with_buckets
                .expect_get()
                .with(eq("RS_BUCKET_1_MAX_BLOCK_SIZE"))
                .return_const(Ok("1MB".to_string()));
            env_with_buckets
                .expect_get()
                .with(eq("RS_BUCKET_1_MAX_BLOCK_RECORDS"))
                .return_const(Ok("1000".to_string()));

            env_with_buckets
                .expect_get()
                .return_const(Err(VarError::NotPresent));

            let cfg = Cfg::from_env(env_with_buckets);
            let components = cfg.build();

            let storage = components.storage.read().await;
            let bucket1 = storage.get_bucket("bucket1").unwrap();

            assert!(bucket1.is_provisioned());
            assert_eq!(bucket1.settings().quota_type, Some(FIFO));
            assert_eq!(bucket1.settings().quota_size, Some(1_000_000_000));
            assert_eq!(bucket1.settings().max_block_size, Some(1_000_000));
            assert_eq!(bucket1.settings().max_block_records, Some(1000));
        }

        #[rstest]
        #[tokio::test]
        async fn test_buckets_defaults(mut env_with_buckets: MockEnvGetter) {
            env_with_buckets
                .expect_get()
                .return_const(Err(VarError::NotPresent));

            let cfg = Cfg::from_env(env_with_buckets);
            let components = cfg.build();

            let storage = components.storage.read().await;
            let bucket1 = storage.get_bucket("bucket1").unwrap();

            assert_eq!(
                bucket1.settings(),
                &Bucket::defaults(),
                "use defaults if env vars are not set"
            );
        }

        #[rstest]
        #[tokio::test]
        async fn test_tokens(mut env_with_tokens: MockEnvGetter) {
            env_with_tokens
                .expect_get()
                .with(eq("RS_TOKEN_1_VALUE"))
                .return_const(Ok("TOKEN".to_string()));
            env_with_tokens
                .expect_get()
                .with(eq("RS_TOKEN_1_FULL_ACCESS"))
                .return_const(Ok("true".to_string()));
            env_with_tokens
                .expect_get()
                .with(eq("RS_TOKEN_1_READ"))
                .return_const(Ok("bucket1,bucket2".to_string()));
            env_with_tokens
                .expect_get()
                .with(eq("RS_TOKEN_1_WRITE"))
                .return_const(Ok("bucket1".to_string()));
            env_with_tokens
                .expect_get()
                .return_const(Err(VarError::NotPresent));

            let cfg = Cfg::from_env(env_with_tokens);
            let components = cfg.build();

            let repo = components.token_repo.read().await;
            let token1 = repo.get_token("token1").unwrap().clone();
            assert_eq!(token1.value, "TOKEN");
            assert!(token1.is_provisioned);

            let permissions = token1.permissions.unwrap();
            assert_eq!(permissions.full_access, true);
            assert_eq!(permissions.read, vec!["bucket1", "bucket2"]);
            assert_eq!(permissions.write, vec!["bucket1"]);
        }

        #[rstest]
        #[tokio::test]
        async fn test_tokens_no_value(mut env_with_tokens: MockEnvGetter) {
            env_with_tokens
                .expect_get()
                .with(eq("RS_TOKEN_1_VALUE"))
                .return_const(Err(VarError::NotPresent));
            env_with_tokens
                .expect_get()
                .return_const(Err(VarError::NotPresent));

            let cfg = Cfg::from_env(env_with_tokens);
            let components = cfg.build();

            let repo = components.token_repo.read().await;
            let err = repo.get_token("token1").err().unwrap();
            assert_eq!(err, ReductError::not_found("Token 'token1' doesn't exist"));
        }
    }
}
